spec-0007/index.md

```diff
@@ -161,62 +177,106 @@ def check_random_state(seed=_NoValue, rng=_NoValue):
                 if dep_version:
-                    end_version = dep_version.split('.')
-                    end_version[1] = str(int(end_version[1]) + 2)
-                    end_version = '.'.join(end_version)
                     message = (f"Use of keyword argument `{old_name}` is "
                                f"deprecated and replaced by `{new_name}`.  "
                                f"Support for `{old_name}` will be removed "
                                f"in SciPy {end_version}.")
                     warnings.warn(message, DeprecationWarning, stacklevel=2)
-                if new_name in kwargs:
-                    message = (f"{fun.__name__}() got multiple values for "
-                               f"argument now known as `{new_name}`")
-                    raise TypeError(message)
-
-            kwargs[new_name] = check_random_state(
-                kwargs.pop(old_name, _NoValue),
-                rng=kwargs.pop(new_name, _NoValue)
-            )
+
+            elif as_pos_arg:  # warn about deprecated positional use
+                # Not required for adoption of this SPEC; feel free to remove
+                args = list(args)
+                args[position_num] = _check_random_state(seed=args[position_num])
+                if dep_version:
+                    message = (f"Positional use of argument `{old_name}` and "
+                               "those that follow is deprecated and will cause "
+                               f"an error in SciPy {end_version}. Replace these "
+                               f"with keyword arguments, and use `{new_name}` "
+                               f"instead of `{old_name}`.")
+                    warnings.warn(message, DeprecationWarning, stacklevel=2)
+
+            elif as_new_kwarg:  # no warnings; this is the preferred use
+                kwargs[new_name] = _check_random_state(rng=kwargs[new_name])
+
+            else:  # will emit FutureWarning if `np.random.seed` was used
+                kwargs[new_name] = _check_random_state()
+
             return fun(*args, **kwargs)
         return wrapper
     return decorator
 
 
-@_prepare_rng("random_state")
-def library_function(/, rng=None):
-    # The decorated library function takes an `rng` argument which is
-    # guaranteed to be a either a Generator or a RandomState.
-    # `random_state=` is supported input (the old can be customized).
+# Example usage of _prepare_rng decorator
+@_prepare_rng("random_state", 1, "1.15.0")
+def library_function(arg1, rng=None, arg2=None):
+    # The decorated library function accepts an `rng` argument, but the user
+    # can also pass a `random_state` argument (as specified by `old_name`) or
+    # a positional argument (with position specified by `position_name`). In
+    # any case, the argument passed to `library_function` will be either a
+    # Generator or a RandomState after validation by the decorator.
```

Comment:
We should specify what happens *after* the deprecation period is over.
If positional use of the PRNG argument is deprecated, then *rng* (and all following arguments) are marked keyword-only.
The decorator gets removed, and `rng = np.random.default_rng(rng)` can be used to validate the argument inside the function?